src/NServiceBus.Callbacks/RequestResponseExtensions.cs

```diff
@@ -0,0 +1,102 @@
+﻿namespace NServiceBus
+{
+    using System;
+    using System.Threading.Tasks;
+    using NServiceBus.Configuration.AdvanceExtensibility;
+
+    /// <summary>
+    /// Synchronous request/response extension methods.
+    /// </summary>
+    public static class RequestResponseExtensions
+    {
+        /// <summary>
+        /// Sends a <paramref name="requestMessage"/> to the configured destination and returns back a <see cref="SendContext{TResponse}"/> that allows the use user to wait on.
+        /// </summary>
+        /// <typeparam name="TResponse">The response type.</typeparam>
+        /// <param name="bus">Object beeing extended.</param>
+        /// <param name="requestMessage">The request message.</param>
+        /// <param name="options">The options for the send.</param>
+        /// <returns>A synchronous request/response context.</returns>
+        public static SendContext<TResponse> SynchronousRequestResponse<TResponse>(this IBus bus, object requestMessage, SynchronousOptions options)
+        {
+            if (requestMessage == null)
+            {
+                throw new ArgumentNullException("requestMessage");
+            }
+
+            if (options == null)
+            {
+                throw new ArgumentNullException("options");
+            }
+
+            if (bus == null)
+            {
+                throw new ArgumentNullException("bus");
+            }
+
+            var sendOptions = new SendOptions(options.Destination, options.CorrelationId);
+                
+            foreach (var header in options.Headers)
+            {
+                sendOptions.AddHeader(header.Key, header.Value);
+            }
+
+            sendOptions
+                .AddHeader("NServiceBus.HasCallback", Boolean.TrueString)
```

Comment:
+1 I like that header better!

I would change to `$Routing.ReplyToSpecificEndpointInstance`

On Thu, Apr 30, 2015 at 10:37 AM, Szymon Pobiega <notifications@github.com>
wrote:

> In src/NServiceBus.Callbacks/RequestResponseExtensions.cs
> <https://github.com/Particular/NServiceBus/pull/2689#discussion_r29411751>
> :
>
> > +            }
> > +
> > +            if (bus == null)
> > +            {
> > +                throw new ArgumentNullException("bus");
> > +            }
> > +
> > +            var sendOptions = new SendOptions(options.Destination, options.CorrelationId);
> > +
> > +            foreach (var header in options.Headers)
> > +            {
> > +                sendOptions.AddHeader(header.Key, header.Value);
> > +            }
> > +
> > +            sendOptions
> > +                .AddHeader("NServiceBus.HasCallback", Boolean.TrueString)
>
> Are we using a proper language here. I mean, is the fact that the message
> has a callback essential or is it any other property? I mean, can there be
> a different use case which should trigger the same routing behavior change
> but is not related to the callbacks. In such case maybe the more
> appropriate name would be $Routing.ReplyToSpecificInstance
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Particular/NServiceBus/pull/2689/files#r29411751>.
>